rpc/src/client/transport/websocket.rs

```diff
@@ -301,35 +301,35 @@ impl WebSocketClientDriver {
     pub async fn run(mut self) -> Result<()> {
         let mut ping_interval =
             tokio::time::interval_at(Instant::now().add(PING_INTERVAL), PING_INTERVAL);
-        let mut recv_timeout = tokio::time::delay_for(PING_INTERVAL);
+
         loop {
+            let next_msg = tokio::time::timeout(RECV_TIMEOUT, self.stream.next());
```

Comment:
@thanethomson This uses `RECV_TIMEOUT` as the timeout for the next message, instead of `PING_INTERVAL`. Can you please confirm that this is correct?

Reply:
@thanethomson I've also had to implement the timeout in a different way, because [`Sleep::reset`](https://docs.rs/tokio/1.0.2/tokio/time/struct.Sleep.html#method.reset) now requires the self-type `Pin<&mut Sleep>` instead of just `&mut Sleep`, and thus takes ownership of the pinned `Sleep` value, which makes it impossible to `.await` it later in the loop.

You can see the error for yourself if you replace the whole `run` method with

```rust
    /// Executes the WebSocket driver, which manages the underlying WebSocket
    /// transport.
    pub async fn run(mut self) -> Result<()> {
        let mut ping_interval =
            tokio::time::interval_at(Instant::now().add(PING_INTERVAL), PING_INTERVAL);

        let recv_timeout = tokio::time::sleep(RECV_TIMEOUT);
        tokio::pin!(recv_timeout);

        loop {
            tokio::select! {
                Some(res) = self.stream.next() => match res {
                    Ok(msg) => {
                        // Reset the receive timeout every time we successfully
                        // receive a message from the remote endpoint.
                        recv_timeout.reset(Instant::now().add(RECV_TIMEOUT));
                        self.handle_incoming_msg(msg).await?
                    },
                    Err(e) => return Err(
                        Error::websocket_error(
                            format!("failed to read from WebSocket connection: {}", e),
                        ),
                    ),
                },
                Some(cmd) = self.cmd_rx.recv() => match cmd {
                    DriverCommand::Subscribe(subs_cmd) => self.subscribe(subs_cmd).await?,
                    DriverCommand::Unsubscribe(unsubs_cmd) => self.unsubscribe(unsubs_cmd).await?,
                    DriverCommand::SimpleRequest(req_cmd) => self.simple_request(req_cmd).await?,
                    DriverCommand::Terminate => return self.close().await,
                },
                _ = ping_interval.tick() => self.ping().await?,
                _ = &mut recv_timeout => {
                    return Err(Error::websocket_error(format!(
                        "reading from WebSocket connection timed out after {} seconds",
                        RECV_TIMEOUT.as_secs()
                    )));
                }
            }
        }
    }
```

```
error[E0382]: borrow of moved value: `recv_timeout`
   --> rpc/src/client/transport/websocket.rs:330:21
    |
306 |         tokio::pin!(recv_timeout);
    |         -------------------------- move occurs because `recv_timeout` has type `std::pin::Pin<&mut tokio::time::Sleep>`, which does not implement the `Copy` trait
...
314 |                         recv_timeout.reset(Instant::now().add(RECV_TIMEOUT));
    |                         ------------ value moved here, in previous iteration of loop
...
330 |                 _ = &mut recv_timeout => {
    |                     ^^^^^^^^^^^^^^^^^ value borrowed here after move

error[E0382]: borrow of moved value: `recv_timeout`
   --> rpc/src/client/transport/websocket.rs:330:21
    |
306 |         tokio::pin!(recv_timeout);
    |         -------------------------- move occurs because `recv_timeout` has type `std::pin::Pin<&mut tokio::time::Sleep>`, which does not implement the `Copy` trait
...
314 |                         recv_timeout.reset(Instant::now().add(RECV_TIMEOUT));
    |                         ------------ value moved here, in previous iteration of loop
...
330 |                 _ = &mut recv_timeout => {
    |                     ^^^^^^^^^^^^^^^^^ value borrowed here after move
```

Please let me know what you think of the new solution with `timeout` combinator :)

Reply:
@thanethomson What do you think about introducing an enum to help with the readability of the client loop:

```rust
    pub async fn run(mut self) -> Result<()> {
        let mut ping_interval =
            tokio::time::interval_at(Instant::now().add(PING_INTERVAL), PING_INTERVAL);

        enum NextMsg {
            Ok(Message),
            StreamEnded,
            Timeout,
            WebSocketError(async_tungstenite::tungstenite::Error),
        }

        loop {
            let next_msg =
                tokio::time::timeout(RECV_TIMEOUT, self.stream.next()).map(|res| match res {
                    Ok(Some(Ok(msg))) => NextMsg::Ok(msg),
                    Ok(Some(Err(e))) => NextMsg::WebSocketError(e),
                    Ok(None) => NextMsg::StreamEnded,
                    Err(_) => NextMsg::Timeout,
                });

            tokio::select! {
                res = next_msg => match res {
                    NextMsg::Ok(msg) => {
                        self.handle_incoming_msg(msg).await?
                    },
                    NextMsg::StreamEnded => {
                        // Websocket stream is over, let's continue in case
                        // we still receive commands via the `cmd_rx` channel.
                        continue;
                    },
                    NextMsg::Timeout => return Err(Error::websocket_error(format!(
                        "reading from WebSocket connection timed out after {} seconds",
                        RECV_TIMEOUT.as_secs()
                    ))),
                    NextMsg::WebSocketError(e) => return Err(Error::websocket_error(
                        format!("failed to read from WebSocket connection: {}", e),
                    )),
                },
                Some(cmd) = self.cmd_rx.recv() => match cmd {
                    DriverCommand::Subscribe(subs_cmd) => self.subscribe(subs_cmd).await?,
                    DriverCommand::Unsubscribe(unsubs_cmd) => self.unsubscribe(unsubs_cmd).await?,
                    DriverCommand::SimpleRequest(req_cmd) => self.simple_request(req_cmd).await?,
                    DriverCommand::Terminate => return self.close().await,
                },
                _ = ping_interval.tick() => self.ping().await?,
            }
        }
    }
```

Reply:
@romac you're right, the `recv_timeout` should've been reset with `RECV_TIMEOUT` and not `PING_INTERVAL`.

But with regard to the timeout combinator, correct me if I'm wrong but it looks like it changes the behaviour of the loop. The expected behaviour is that the `recv_timeout` must only be reset when we receive a message.

The new behaviour seems to reset the timeout on each iteration of the loop, which would also then happen when we (1) receive any command, and (2) send out a ping.

Reply:
> But with regard to the timeout combinator, correct me if I'm wrong but it looks like it changes the behaviour of the loop. The expected behaviour is that the recv_timeout must only be reset when we receive a message.

Yeah absolutely, I understood the original semantics but somehow didn't realize that this would change the behavior.

Thankfully, I managed to restore the original behavior in 477881026205da4d8911c5e59a9bccd9e5e55f15. They key was to use [`Pin::as_mut`](https://doc.rust-lang.org/stable/std/pin/struct.Pin.html#method.as_mut) to reborrow the pinned `&mut Sleep` in order to call `reset` on it.